### PR TITLE
NIFI-2850: Added a migrate() method to ProcessSession and refactored …

### PR DESCRIPTION
**Pull request by markap14**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:
### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
   in the commit message?
- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?
- [ ] Is your initial contribution a single, squashed commit?
### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?
### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
### Note:

Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

…BinFiles and MergeContent to use it

### CODE REVIEW

**Comment by olegz** [comment]
Reviewing. . .

File: nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/bin/BinFiles.java [hunk 1]
```diff
@@ -273,25 +262,26 @@ private int binFlowFiles(final ProcessContext context, final ProcessSessionFacto
...
-            // could not be added to a bin -- probably too large by itself, so create a separate bin for just this guy.
-            if (!binned) {
-                Bin bin = new Bin(0, Long.MAX_VALUE, 0, Integer.MAX_VALUE, null);
-                bin.offer(flowFile, session);
-                this.readyBins.add(bin);
+            final Map<String, List<FlowFile>> flowFileGroups = new HashMap<>();
+            for (FlowFile flowFile : flowFiles) {
+                flowFile = this.preprocessFlowFile(context, session, flowFile);
+                final String groupingIdentifier = getGroupId(context, flowFile);
+                flowFileGroups.computeIfAbsent(groupingIdentifier, id -> new ArrayList<>()).add(flowFile);
```
Comment (markap14) [inline nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/bin/BinFiles.java]: I don't believe so. Using putIfAbsent, we would be creating a new ArrayList every time. By using the computeIfAbsent, it allows us to create the ArrayList only if the key is not already present.

File: nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/bin/BinFiles.java [hunk 2]
```diff
@@ -273,25 +262,26 @@ private int binFlowFiles(final ProcessContext context, final ProcessSessionFacto
...
 
-            flowFilesBinned++;
+            for (final Map.Entry<String, List<FlowFile>> entry : flowFileGroups.entrySet()) {
+                final Set<FlowFile> unbinned = binManager.offer(entry.getKey(), entry.getValue(), session, sessionFactory);
+                for (final FlowFile flowFile : unbinned) {
+                    Bin bin = new Bin(session, 0, Long.MAX_VALUE, 0, Integer.MAX_VALUE, null);
+                    bin.offer(flowFile, session);
+                    this.readyBins.add(bin);
+                }
+            }
```
Comment (markap14) [inline nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/bin/BinFiles.java]: Not exactly. The loop above says "if the bin manager didn't bin it for whatever reason, create our own one-element bin and process it (by adding to this.readyBins) - nothing else will go in this bin." In BinManager:201, it is saying "if the FlowFile didn't fit in any of the bins that are available, create a new bin and add this FlowFile to it. Subsequent FlowFiles may then go into this bin."

**Comment by markap14** [comment]
@olegz Thanks for reviewing! I have responded to your concerns above. If you feel they need more discussion, then we certainly can. Thanks!

**Comment by olegz** [comment]
@markap14 all is good. +1 here but think someone else should give it a look as well given its complexity.

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveStreaming.java [hunk 3]
```diff
@@ -377,7 +377,7 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
         final AtomicInteger successfulRecordCount = new AtomicInteger(0);
         List<HiveStreamingRecord> successfulRecords = new LinkedList<>();
         final FlowFile inputFlowFile = flowFile;
-        final AtomicBoolean incomingFlowFileTransferred = new AtomicBoolean(false);
+        final AtomicBoolean processingFailure = new AtomicBoolean(false);
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveStreaming.java]: why is there any change to the hive streaming processor in this commit/PR?

Comment (markap14) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveStreaming.java]: @joewitt so in updating MockProcessSession to support the session migration, I found that the MockProcessSession was not properly keeping track of the 'recursionSet' that StandardProcessSession was. I fixed that bug, as it had to be addressed in order to properly implement the session migration in the Mock framework. Once I fixed that bug in Mock ProcessSession, it then exposed this unrelated bug in PutHiveStreaming. However, the bug in PutHiveStreaming would now cause unit test failures, I addressed that in this PR as well.

**Pull request closed without merging**